### PR TITLE
[Lock] MongoDbStore skim non-standard options from uri

### LINKED ISSUE #37180: [Lock] remove non-standard MongoDbStore URI parameters before passing to driver
**Symfony version(s) affected**: 5.1.0

**Description**  
@jmikola  has raised concerns with our handling of both the `database` path and `collection` query string in a `MongoDbStore` constructed with a MongoDB connection URI. Please see https://github.com/symfony/symfony/pull/37140#issuecomment-641665443

**How to reproduce**  
Construction of a `MongoDbStore` using a MongoDB connection URI with `collection=` in the querystring.

**Possible Solution**  
We should remove `collection` from the query string before passing the URI to the driver.

**Additional context**  
https://github.com/symfony/symfony/pull/37140#issuecomment-641665443

See the note regarding libmongoc client and connection re-use internally: https://www.php.net/manual/en/mongodb.connection-handling.php

### PR DESCRIPTION
**Pull request by kralos**
| Q             | A
| ------------- | ---
| Branch?       | 5.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #37180
| License       | MIT
| Doc PR        | 

Don't allow non-standard connection uri options reach `MongoDB\Client`

### CODE REVIEW

File: src/Symfony/Component/Lock/Store/MongoDbStore.php [hunk 1]
```diff
@@ -143,6 +129,41 @@ public function __construct($mongo, array $options = [], float $initialTtl = 300
...
+        if (false === $parsedUrl = parse_url($uri)) {
+            throw new InvalidArgumentException(sprintf('The given MongoDB Connection URI "%s" is invalid.', $mongo));
+        }
+
+        $query = [];
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $query);
+        }
+
+        if (null === ($this->options['collection'] ?? null) && isset($query['collection'])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Lock/Store/MongoDbStore.php:146]: I know this was already the case before your PR, but the precedence looks wrong to me:
the DSN should have higher precedence over options (I just double-checked, that's how the Cache component does also).

/cc @jderusse maybe we should review all DSN parsing logic to ensure the precedence is consistent everywhere
/cc @Nyholm that's a good case for [the DSN component](https://github.com/symfony/symfony/pull/36999) - formalizing this behavior :)

Comment (jderusse) [reply src/Symfony/Component/Lock/Store/MongoDbStore.php:146]: :+1: not only in cache and lock. 

messenger Doctrine:
```php
$configuration += $options + $query + static::DEFAULT_OPTIONS;
```
messenger Redis:
```php
parse_str($parsedUrl['query'], $options);
```
messenger Sqs
```php
 $configuration = ['buffer_size'] = $options['buffer_size'] ?? (int) ($query['buffer_size'] ?? self::DEFAULT_OPTIONS['buffer_size']),
```
messenger amqp
```php
$amqpOptions = array_replace_recursive([], $options, $parsedQuery);
```

Comment (nicolas-grekas) [reply src/Symfony/Component/Lock/Store/MongoDbStore.php:146]: Bad news, we need to fix this. Up for a PR?

Comment (kralos) [reply src/Symfony/Component/Lock/Store/MongoDbStore.php:146]: @nicolas-grekas In order to change the precedence, I think we should detect `dsn` / `options` where a precedence change would cause an effective option change and instead `trigger_deprecation` for now. Then swap the precedence in `symfony` `6.0`. Do you want me to add this detection / `trigger_deprecation` in this PR for `Lock\Store\MongoDbStore` only or should we make a list of ALL `dsn` / `options` precedence affected classes and do them in a separate PR?

Comment (kralos) [reply src/Symfony/Component/Lock/Store/MongoDbStore.php]: I've added the precedence change since you guys seem to be classing this as a bug rather than a deprecation in: https://github.com/symfony/symfony/pull/37268

Comment (nicolas-grekas) [reply src/Symfony/Component/Lock/Store/MongoDbStore.php]: There is no way to deprecate this behavior properly. Also this will very unlikely hit anyone. This precedence issue lessens the capabilities of the ops side of things. Basically, having the options take over the DSN means that the source needs to change (the options) instead of some env vars. That's a management issue at this point, and this change fixes a process issue: the env vars should always win to give control to the ops, which are the ones that know better since they're the ones in touch with the effective infra.

TL;DR, this is a bugfix.

**Comment by nicolas-grekas** [comment]
Can you please rebase now that the priority is fixed?

File: src/Symfony/Component/Lock/Store/MongoDbStore.php [hunk 2]
```diff
@@ -143,6 +129,41 @@ public function __construct($mongo, array $options = [], float $initialTtl = 300
         }
     }
 
+    /**
+     * Extract default database and collection from given connection uri and remove collection querystring.
+     */
+    private function skimUri(string $uri): string
+    {
+        if (false === $parsedUrl = parse_url($uri)) {
+            throw new InvalidArgumentException(sprintf('The given MongoDB Connection URI "%s" is invalid.', $mongo));
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/MongoDbStore.php:138]: $mongo is not defined

File: src/Symfony/Component/Lock/Store/MongoDbStore.php [hunk 3]
```diff
@@ -143,6 +129,41 @@ public function __construct($mongo, array $options = [], float $initialTtl = 300
         }
     }
 
+    /**
+     * Extract default database and collection from given connection uri and remove collection querystring.
+     */
+    private function skimUri(string $uri): string
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/MongoDbStore.php:135]: I'm not comfortable with this method, a lot of complex logic to extract parameter from URI mixing parse_str and strpos.

Givent the format of the initial `DSN` is defined by symfony, wouldn't it be safer to build the mongo URI instead?

**Comment by ck-developer** [comment]
hello I would like to share some use cases on precedence of DSN options

case: 1
.env.dist
```
MESSENGER_TRANSPORT_DSN=sqs://localhost
```
config/packages/messenger.yaml
```
framework:
    messenger:
        transports:
            queue_1:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    queue_name: 'queue_1'
            queue_2:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    queue_name: 'queue_2'
```

in this case when I setup transport I have empty queue names.

case: 2
I think that this problem is also linked to the priority of the dsn https://github.com/symfony/symfony/issues/37293

File: src/Symfony/Component/Lock/Store/MongoDbStore.php (hunk 2 above)
Comment (kralos) [reply src/Symfony/Component/Lock/Store/MongoDbStore.php:138]: fixed

File: src/Symfony/Component/Lock/Store/MongoDbStore.php (hunk 3 above)
Comment (kralos) [reply src/Symfony/Component/Lock/Store/MongoDbStore.php:135]: the reason i chose to use strpos to re-assemble was to simplify the re-assembly. there is no reverse of parse_str in php. see https://www.php.net/manual/en/function.parse-url.php#106731

I'm trying not to get too involved with touching the URI. What would you suggest that is simpler?

**Comment by jderusse** [comment]
@nicolas-grekas LGTM from a lock point of view. I just have concerns about complexity of url generator https://github.com/symfony/symfony/pull/37218#discussion_r442759940 . If you think it's ok you can merge it.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @kralos.

File: src/Symfony/Component/Lock/Store/MongoDbStore.php [hunk 4]
```diff
@@ -143,6 +129,45 @@ public function __construct($mongo, array $options = [], float $initialTtl = 300
...
+            $this->options['collection'] = $query['collection'];
+            $queryStringPos = strrpos($uri, $parsedUrl['query']);
+            unset($query['collection']);
+            $prefix = substr($uri, 0, $queryStringPos);
+            $newQuery = http_build_query($query, '', '&', PHP_QUERY_RFC3986);
+            if (empty($newQuery)) {
+                $prefix = rtrim($prefix, '?');
+            }
+            $suffix = substr($uri, $queryStringPos + \strlen($parsedUrl['query']));
+            $uri = $prefix.$newQuery.$suffix;
```
Comment (jmikola) [inline src/Symfony/Component/Lock/Store/MongoDbStore.php:160]: @kralos: I only just came across this PR based on a notification in #37180. I understand what this function is attempting to do, but using `parse_str` may silently corrupting a connection string with repeated `readPreferenceTags` keys (a valid use case, as explained in the [URI options](https://github.com/mongodb/specifications/blob/master/source/uri-options/uri-options.rst) spec). [This comment](https://www.php.net/manual/en/function.parse-str.php#76792) on PHP.net goes into more detail about that.

I think it would be preferable to capture the URI option with a regular expression and then, if anything was found, strip it from the returned string. While collection names have their [own restrictions](https://docs.mongodb.com/manual/reference/limits/#Restriction-on-Collection-Names), for purposes of URI parsing I think it'd be best to use something like `/collection=([^&]*)/i`.

Comment (kralos) [reply src/Symfony/Component/Lock/Store/MongoDbStore.php:160]: see https://github.com/symfony/symfony/issues/37864 / https://github.com/symfony/symfony/pull/37868
